File: python/1653. Minimum Deletions to Make String Balanced.py
class Solution(object):
    def minimumDeletions(self, s):
        """
        :type s: str
        :rtype: int
        """
        n = len(s)
        dp = [[0]*2 for _ in range(n+1)]
        for i in range(1, n+1):
            if s[i-1]=='a':
                dp[i][0] = dp[i-1][0]
                dp[i][1] = min(dp[i-1][0], dp[i-1][1]) + 1
            else:
                dp[i][0] = dp[i-1][0] + 1
                dp[i][1] = min(dp[i-1][0], dp[i-1][1])
        return min(dp[-1])

# 思路：dp[i][0]表示前i个字符，以'a'结尾的平衡字符串的最少删除次数
#       dp[i][1]表示前i个字符，以'b'结尾的平衡字符串的最少删除次数

File: go/src/leetcode/1653/1653. Minimum Deletions to Make String Balanced.go
package main

import "fmt"

/*
*
https://leetcode.cn/problems/minimum-deletions-to-make-string-balanced/
baab
bbab
*/
func minimumDeletions(s string) int {
	n := len(s)
	left, right := make([]int, n), make([]int, n)
	var cb, ca int
	for i := range s {
		left[i] = cb
		if s[i] == 'b' {
			cb++
		}
	}
	for i := n - 1; i >= 0; i-- {
		right[i] = ca
		if s[i] == 'a' {
			ca++
		}
	}

	var ans = n
	for i := 0; i < n; i++ {
		ans = min(ans, left[i]+right[i])
	}

	return ans
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func main() {
	fmt.Println(minimumDeletions("aababbab"))
	fmt.Println(minimumDeletions("bbaaaaabb"))
}
